File: leetcoding/medium-problems/medium.1371/medium.1371.py
class Solution:
  @classmethod
  def findTheLongestSubstring(self, s: str) -> int:
    vowels = {'a': 1 << 0, 'e': 1 << 1, 'i': 1 << 2, 'o': 1 << 3, 'u': 1 << 4}
    first_occurrence = {0: -1}
    mask = 0
    longest = 0

    for i, char in enumerate(s):
        if char in vowels:
            mask ^= vowels[char]
        if mask in first_occurrence:
            longest = max(longest, i - first_occurrence[mask])
        else:
            first_occurrence[mask] = i

    return longest

print(Solution.findTheLongestSubstring("eleetminicoworoep"))  
print(Solution.findTheLongestSubstring("leetcodeisgreat"))
print(Solution.findTheLongestSubstring("bcbcbc"))

File: leetcoding/medium-problems/medium.1371/medium.1371.go
package main

import (
	"fmt"
)

func findTheLongestSubstring(s string) int {
	firstOccurrence := make(map[int]int)
	firstOccurrence[0] = -1 
	mask := 0    
	longest := 0 

	vowels := map[byte]int{
		'a': 1 << 0,
		'e': 1 << 1,
		'i': 1 << 2,
		'o': 1 << 3,
		'u': 1 << 4,
	}

	for i := 0; i < len(s); i++ {
		char := s[i]

		if val, ok := vowels[char]; ok {
			mask ^= val
		}

		if pos, found := firstOccurrence[mask]; found {
			longest = max(longest, i-pos)
		} else {
			firstOccurrence[mask] = i
		}
	}

	return longest
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func main() {

	fmt.Println(findTheLongestSubstring("eleetminicoworoep")) 
	fmt.Println(findTheLongestSubstring("leetcodeisgreat"))   
	fmt.Println(findTheLongestSubstring("bcbcbc"))            
}